### ISSUE TITLE
Some issues occurred while executing the triviaqa.sh script on a single-node, 8-GPU

### ISSUE THREAD

**Issue by Boom-1995** [pat-jj/DeepRetrieval#31]
Thank you for your open-sourcing work! I'm trying to reproduce your results on the **NQ task**.
After some debugging, I managed to successfully start training on the **TriviaQA dataset** using the **triviaqa.sh** script, based on **1M Wikipedia passages**. However, I encountered a few issues that I'd like to ask for your help.
First, in terms of metrics, I noticed some discrepancies:

1. The values of **actor/ppo_kl** and **actor/pg_clipfrac** remain at zero throughout training, which is inconsistent with what I saw in your wandb logs. After some investigation, I found that this seems to be caused by old_log_prob and log_prob always being the same. I suspect this might be due to the actor and critic models sharing the same weights—is that not allowed in your setup?
2. Secondly, during the first epoch of training, the reward fluctuates significantly. However, as soon as the second epoch begins, the reward quickly drops to zero and stays there.

Here is my launch script and the wandb logs. Could you please help me analyze them? Thank you again for your help!

**triviaqa.sh**
/home/.conda/envs/deepretrieval/bin/python /DeepRetrieval/code/verl/trainer/main_ppo.py \
    data.train_files=/DeepRetrieval/train_data/triviaqa/train.parquet \
    data.val_files=/DeepRetrieval/train_data/triviaqa/val.parquet \
    data.train_batch_size=16 \
    data.val_batch_size=16 \
    data.max_prompt_length=256 \
    data.max_response_length=350 \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.strategy=fsdp \
    actor_rollout_ref.actor.ppo_mini_batch_size=16 \
    actor_rollout_ref.actor.ppo_micro_batch_size=8 \
    critic.ppo_micro_batch_size=8\
    actor_rollout_ref.rollout.log_prob_micro_batch_size=8\
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.8 \
    actor_rollout_ref.ref.log_prob_micro_batch_size=8 \
    actor_rollout_ref.rollout.temperature=0.6 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    critic.model.enable_gradient_checkpointing=True \
    critic.optim.lr=1e-5 \
    critic.model.enable_gradient_checkpointing=True \
    algorithm.kl_ctrl.kl_coef=0.001 \
    trainer.logger=['wandb'] \
    +trainer.val_before_train=False \
    trainer.resume_from_checkpoint=true \
    trainer.default_hdfs_dir=null \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=500 \
    trainer.test_freq=500 \
    trainer.project_name=$PROJECT_NAME \
    trainer.experiment_name=$EXP_NAME \
  actor_rollout_ref.model.path=/open_source_llm/Qwen2.5-3B-Instruct \
  critic.model.path=/open_source_llm/Qwen2.5-3B-Instruct \
  trainer.default_local_dir=/model/DeepRetrieval/Qwen2.5-3B \
  trainer.total_epochs=5 2>&1 | tee exp_log/$PROJECT_NAME-3b-ppo-verl_demo_$DATE.log 

**Here is my wandb log**

When the **global_step** exceeds around 4.5k, it indicates that the training has entered the second epoch.

**actor**
![Image](https://github.com/user-attachments/assets/628cfcb3-a1bf-4dd3-8ff2-37e216586ede)

**reward**

![Image](https://github.com/user-attachments/assets/2526ca79-5c15-47f6-99dc-52edf6e0a233)

**response_length**

![Image](https://github.com/user-attachments/assets/9b784571-a5fe-449d-b673-1e369199f893)

**Comment by pat-jj** [comment]
Hi, thanks for your question! I saw `trainer.resume_from_checkpoint=true` in your script - are you using the same version of verl as ours? Also, have you tried to use two GPUs and apply our settings [here](https://github.com/pat-jj/DeepRetrieval/blob/main/code/scripts/train/triviaqa.sh)?

**Comment by Boom-1995** [comment]
Thank you for your prompt response!!
First, I would like to explain the parameter **trainer.resume_from_checkpoint=true**. The GPU cluster I am using operates under a scheduling mechanism, which prevents me from completing the entire training process in one go. As a result, I need to implement a checkpointing mechanism in the code, so that the training task can resume from an intermediate state when it restarts next time. Therefore, the training code and data I use are largely aligned with those in your repository, with the only modification being the addition of checkpointing support to enable resuming from interruptions.

Thank you again for your suggestion. I tried running the experiment using two GPUs, but observed that the **ppo_kl** and  **pg_clipfrac** value still remained consistently at 0.

In addition, I conducted another experiment using a single GPU with the Qwen-0.5B model. Interestingly, in this case, the ppo_kl metric changes as the training step increases.
I'm curious—why does this difference occur?

Below is the WandB log from the single-GPU training run. Even after just a few steps, noticeable changes in the metrics can already be observed.

![Image](https://github.com/user-attachments/assets/327975af-818c-4fd0-a6a4-faf8df75282e)

**Comment by pat-jj** [comment]
This is an interesting observation. We’ll look into it more deeply in a few days, as our team is currently quite busy with the NeurIPS and EMNLP deadlines. The only suggestion I can give now is to create a new folder, re-download our repo, and run the exact same code. If that doesn’t work, I’ll try to reproduce the issue using your setup.

**Comment by Boom-1995** [comment]
Thank you for your suggestions! After several more experiments, I successfully resolved this issue. I found that the problem described in this issue was related to the parameter settings in the startup script. Although I haven't fully understood why this behavior occurs, it is clearly directly associated with the parameters **train_batch_size**, **ppo_mini_batch_size**, and **ppo_micro_batch_size**.

On another note, I tried comparing the model I trained with your open-source model, **DeepRetrieval/DeepRetrieval-TriviaQA-BM25-3B**, but the performance is significantly worse than yours. (So far, I've trained my model for 3 epochs, but the evaluation results for the last two epochs are nearly identical to those of the first epoch.) Below is a comparison of the results on the **TriviaQA test set**.

![Image](https://github.com/user-attachments/assets/4811ab83-1326-4e9d-828f-27ef54ae92a1)

I noticed that in your script, the paths for the models are set as follows:
**actor_rollout_ref.model.path=/shared/eng/pj20/lmr_model/nq_serini_3b/actor/global_step_1800** and
**critic.model.path=/shared/eng/pj20/lmr_model/nq_serini_3b/critic/global_step_1800**.
However, since I couldn't locate these specific models, I used the **Qwen-2.5-3B-Instruct model** as a substitute.

Could this be the reason why I wasn't able to reproduce the performance of your released model?
Or are there other details I might have overlooked?
Below are my reward training curves and actor-related metrics 

![Image](https://github.com/user-attachments/assets/a7c80a8d-4f5b-4f13-915c-09d9ba893ac0)

![Image](https://github.com/user-attachments/assets/9b9e868e-9816-4c44-bf4f-52d939578931)

Thank you again for your kind and helpful responses!

**Comment by pat-jj** [comment]
Hi, glad to hear you were able to resolve the training issue!

Regarding your new question:
We encountered a similar problem during our own training. For TriviaQA, starting from scratch with Qwen2.5-3B-Instruct often led to instability and convergence to local minima. However, when we initialized training from a model already fine-tuned on the NQ dataset, the process became much more stable.

I’d recommend starting from our NQ-pretrained model here: [DeepRetrieval-NQ-BM25-3B](https://huggingface.co/DeepRetrieval/DeepRetrieval-NQ-BM25-3B) for finetuning on TriviaQA.

BTW, if you're interested on the RAG task, feel free check our new work [s3](https://arxiv.org/pdf/2505.14146), which is a more advanced framework for training search agents for QA.
